core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```
@@ -140,79 +141,154 @@ class TransactionStateManager(brokerId: Int,
-                      && txnMetadata.producerEpoch == idCoordinatorEpochAndMetadata.transitMetadata.producerEpoch
-                      && response.error == Errors.NONE) {
-                      txnMetadataCacheEntry.metadataPerTransactionalId.remove(transactionalId)
-                    } else {
-                      warn(s"Failed to remove expired transactionalId: $transactionalId" +
-                        s" from cache. Tombstone append error code: ${response.error}," +
-                        s" pendingState: ${txnMetadata.pendingState}, producerEpoch: ${txnMetadata.producerEpoch}," +
-                        s" expected producerEpoch: ${idCoordinatorEpochAndMetadata.transitMetadata.producerEpoch}," +
-                        s" coordinatorEpoch: ${txnMetadataCacheEntry.coordinatorEpoch}, expected coordinatorEpoch: " +
-                        s"${idCoordinatorEpochAndMetadata.coordinatorEpoch}")
-                      txnMetadata.pendingState = None
-                    }
+        case None =>
+          (Seq.empty, MemoryRecords.EMPTY)
+      }
+    }
+  }
+
+  private def shouldExpire(
+    txnMetadata: TransactionMetadata,
+    currentTimeMs: Long
+  ): Boolean = {
+    val isExpirableState = txnMetadata.state match {
+      case Empty | CompleteCommit | CompleteAbort => true
+      case _ => false
+    }
+
+    isExpirableState && txnMetadata.txnLastUpdateTimestamp <= currentTimeMs - config.transactionalIdExpirationMs
+  }
+
+  private def maybeAppendExpiration(
+    txnMetadata: TransactionMetadata,
+    recordsBuilder: MemoryRecordsBuilder,
+    currentTimeMs: Long,
+    maxBatchSize: Int
+  ): Boolean = {
+    val keyBytes = TransactionLog.keyToBytes(txnMetadata.transactionalId)
+    if (recordsBuilder.hasRoomFor(currentTimeMs, keyBytes, null, Record.EMPTY_HEADERS)) {
+      recordsBuilder.append(currentTimeMs, keyBytes, null, Record.EMPTY_HEADERS)
+      true
+    } else {
+      if (recordsBuilder.numRecords == 0) {
```

It seems to me that this condition will never be true because `hasRoomFor` returns `true` if `numRecords == 0` so the first record will always be appended. Were you trying to warn if we can't even append the first record to the batch?

---

Yeah, I guess this is the downside of using something like `MemoryRecordsBuilder` which is so tailored to the producer. Would it make sense to add a stricter `hasRoomFor`?

---

Yeah, why not.

---

I decided to get rid of this check instead. It's pretty unlikely that a user would tune max.message.bytes to a small enough value that not even a single record can fit, but if they do so, we can let the write fail with the MESSAGE_TOO_LARGE error that happens today.